src/Network/Transport/TCP.hs

```diff
@@ -1450,7 +1452,7 @@ schedule theirEndPoint act = do
 -- 'runScheduledAction' or by another).
 runScheduledAction :: EndPointPair -> Action a -> IO a
 runScheduledAction (ourEndPoint, theirEndPoint) mvar = do
-    join $ readChan (remoteScheduled theirEndPoint)
+    join $ uninterruptibleMask_ $ readChan (remoteScheduled theirEndPoint)
```

Comment:
> Can this line block for an arbitrarily long time?

No, this is not a arbitrary channel. This channel designed to have a invariant that there is a 1 read after 1 write. So if process tries to read channel it's guaranteed that there is a value in channel.
If by any reason this invariant does not hold then we have a deeper problem in code. However this PR fixes few problems with invariant invalidation.

As @facundominguez mentioned and I showed above it's `readChan` is interruptible operation if there is more then 1 concurrent reader, even if channel is not empty. This happens because there is a read lock `MVar` and read of that `MVar` is interruptible if any other process take a lock.